src/AutoRest.CSharp/Common/Input/Source/ProtocolCompilationInput.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using AutoRest.CSharp.AutoRest.Plugins;
+using AutoRest.CSharp.Generation.Types;
+using Microsoft.CodeAnalysis;
+
+namespace AutoRest.CSharp.Input.Source
+{
+    public class ProtocolCompilationInput : CompilationInput
+    {
+        private List<IMethodSymbol>? _methodSet;
+
+        public static async Task<CompilationInput?> TryCreate(string? existingProjectFolder)
+        {
+            return Configuration.ExistingProjectFolder != null ?
+                new ProtocolCompilationInput(await GeneratedCodeWorkspace.CreateExistingCodeProject(Configuration.ExistingProjectFolder).GetCompilationAsync()) : null;
+        }
+
+        private ProtocolCompilationInput(Compilation compilation)
+            : base(compilation) { }
+
+        private protected override void FilterSymbols()
+        {
+            _methodSet = new List<IMethodSymbol>();
+            foreach (IModuleSymbol module in _compilation.Assembly.Modules)
+            {
+                foreach (var type in SourceInputHelper.GetSymbols(module.GlobalNamespace))
+                {
+                    if (type is INamedTypeSymbol typeSymbol && IsClient(typeSymbol))
+                    {
+                        foreach (var member in typeSymbol.GetMembers())
+                        {
+                            if (member is IMethodSymbol methodSymbol && IsProtocolMethod(methodSymbol))
+                            {
+                                _methodSet.Add(methodSymbol);
+                            }
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
we could use our common pattern to write this:
```suggestion
        private List<IMethodSymbol>? _methodSet;
        private List<IMethodSymbol> MethodSet => _methodSet ??= EnsureMethodSet();
        private List<IMethodSymbol> void FilterSymbols()
        {
            var result = new List<IMethodSymbol>();
            foreach (IModuleSymbol module in _compilation.Assembly.Modules)
            {
                foreach (var type in SourceInputHelper.GetSymbols(module.GlobalNamespace))
                {
                    if (type is INamedTypeSymbol typeSymbol && IsClient(typeSymbol))
                    {
                        foreach (var member in typeSymbol.GetMembers())
                        {
                            if (member is IMethodSymbol methodSymbol && IsProtocolMethod(methodSymbol))
                            {
                                result.Add(methodSymbol);
                            }
                        }
                    }
                }
            }
            return result;
        }
```